MungPlex/netplay/WebsocketClient.cpp

```diff
@@ -0,0 +1,364 @@
+	if (PasswordLength < 2) {
+		MungPlex::Log::LogInformation(MungPlex::LogMessages::NetplayPasswordTooShort);
+		return;
+	}
+	std::vector<uint8_t> BufferToSend(1 + PasswordLength);
+	BufferToSend[0] = HOST_GAME_REQUEST;
+	memcpy(&BufferToSend[1], Password, PasswordLength);
+	GetInstance().SendBinaryData(BufferToSend);
+}
+
+void MungPlex::WebsocketClient::JoinGame(char* _GameID, char* _Password) {
+	if (!GetInstance().IsConnected) {
+		MungPlex::Log::LogInformation(MungPlex::LogMessages::NetplayJoinGameAttemptDisconnected);
+		return;
+	}
+	if (GetInstance().IsInGame) {
+		MungPlex::Log::LogInformation(MungPlex::LogMessages::NetplayJoinGameAttemptAlreadyInOne);
+		return;
+	}
+	if (GetInstance().IsHosting) {
+		MungPlex::Log::LogInformation(MungPlex::LogMessages::NetplayJoinGameAttemptButHosting);
+		return;
+	}
+	uint8_t PasswordLength = static_cast<uint8_t>(strlen(_Password));
+	// The _GameID is always 4 characters, so we don't have to calculate it's length.  Theoritically, we could get the game ID down to only 2 bytes, since the max unsigned 2 byte integer is ~65000, but join game requests don't happen often, so this optimization can wait.
+	std::vector<uint8_t> BufferToSend(5 + PasswordLength);
+	BufferToSend[0] = JOIN_GAME_REQUEST;
+	memcpy(&BufferToSend[1], _GameID, 4);
+	memcpy(&BufferToSend[5], _Password, PasswordLength);
+	GetInstance().GameID = _GameID;
+	GetInstance().GamePassword = _Password;
+	GetInstance().SendBinaryData(BufferToSend);
+}
+
+void MungPlex::WebsocketClient::Disconnect() {
+	if (!GetInstance().IsConnected) {
+		MungPlex::Log::LogInformation(MungPlex::LogMessages::NetplayDisconnectAttemptButAlreadyDisconnected);
+		return;
+	}
+
+	try {
+		GetInstance().Websocket.close(boost::beast::websocket::close_code::normal);
+		GetInstance().IsConnected = false;
+		GetInstance().IsHosting = false;
+		GetInstance().IsInGame = false;
+		MungPlex::Log::LogInformation(MungPlex::LogMessages::NetplaySuccessfullyConnectedToServer);
+	}
+	catch (const std::exception& e) {
+		MungPlex::Log::LogInformation(MungPlex::LogMessages::NetplayDisconnectAttemptButAlreadyDisconnected, e.what());
+	}
+}
+void MungPlex::WebsocketClient::LeaveGame() {
+	try {
+		if (!GetInstance().IsInGame) {
+			MungPlex::Log::LogInformation(MungPlex::LogMessages::NetplayLeaveGameInGame);
+			return;
+		}
+		if (!GetInstance().IsConnected) {
+			MungPlex::Log::LogInformation(MungPlex::LogMessages::NetplayLeaveGameDisconnected);
+			return;
+		}
+		if (GetInstance().IsHosting) {
+			MungPlex::Log::LogInformation(MungPlex::LogMessages::NetplayLeaveGameHosting);
+			return;
+		}
+
+		GetInstance().SendRequestFromEnum(MungPlex::WebsocketClient::DISCONNECT_FROM_GAME_REQUEST);
+	}
+	catch (const std::exception& e) {
+		MungPlex::Log::LogInformation(MungPlex::LogMessages::NetplayDisconnectAttemptButAlreadyDisconnected, e.what());
+	}
+}
+
+void MungPlex::WebsocketClient::SendBinaryData(std::vector<uint8_t> _buffer) {
```

Comment:
It's likely nicer and cleaner to exchange JSON files instead of buffers. In the long term you'll probably be better off, especially for debugging purposes. `nlohmann-json` is already part of the project.

Reply:
Does it really make such a difference? You could apply zlib compression on top to save network resources as well 😅
Boost zlib compression example: https://stackoverflow.com/a/65911734

Reply:
I’ll switch them all to json too I guess, thanks for all the feedback and I’ll update the code as soon as possible!